### PR TITLE
Windows build

### PR DESCRIPTION
**Pull request by jktjkt**
# Building on Windows under MSVC
This patch series makes it possible to build libyang on Windows using MSVC. Important changes which I had to make:

- dynamic loading of plugins is not supported (check the commit description below)
- all dated are considered as "invalid timezone" for purposes of canonization (`struct tm` has no TZ info on Windows)
- there's no interactive shell in `yanglint`, and similarly there's no `yangre` at all due to missing libraries
- multi-dir module search path is specified via `;` due to Windows path format

The rest is the usual porting churn, compatibility functions, CRLF line ends, backslashes in paths, the usual fun. I tried to limit myself to `compat/`, but there's a bunch of `#ifdef`s where it made better sense.

---
## tests: do not hardcode paths under `/tmp`

This is a classic case of insecure file creation, aka CWE-377 (albeit "just" affecting developers, not users of libyang). This path is also not available on Windows.

## include: make self-sufficient

This was using the `LY_ERR` which lives in `log.h` without including anything.

## Do not try to compare strings using C preprocessor

This code apparently tried to feed a CMake-level string into a C preprocessor for comparison via `==`, which won't work. For example, it evaluates to true on MSVC. Fix this by checking compiler-specific macros.

## Allow building on platforms without `struct tm.tm_gmtoff`, `timezone`, `daylight`

I'm not happy about this because the results are non-intuitive, but that looks like something which originates in libyang's attempts to use local TZ everywhere instead of normalizing to UTC time.

But hey, this only matters on a platform which cannot "easily" represent timezones in `struct tm`, or for which we cannot determine the TZ/DST status without going the extra mile of writing non-POSIX stuff. So as long as we mark the returned value with "I cannot really vouch for the TZ", this *might* even be standard-compliant.

## MSVC: handle symbol visibility

On MSVC, the exported symbols have to be marked both in the header and in the implementation as `dllexport`, whereas on GCC the `-fvisibility=normal` attribute is only supposed to be used in the implementation. As a result, we need two types of macros here, one for the public headers, and the other for TUs with implementation. Also, the symbol name cannot be generic ("`API`") because we only want to mark as dllexport those symbols which are in a library that we're building *now*. Otherwise this will break havoc in any libraries which also use the `ABI` macro *and* also use libyang.

## MSVC: POSIX compatibility files

There's a bunch of packages on vcpkg, let's use them this way. I wasn't able to find anything for `<unistd.h>`, so I took that from https://stackoverflow.com/a/826027/2245623 .

## compat: `<unistd.h>`: make `X_OK` a no-op on Windows

There's nothing to check via `access()` on Windows for directory permissions, and I have no clue if the underlying `_access()` ignores unknown mode flags, so let's play it safe. The worst case is that `access()` will be too optimistic, which wouldn't really hart because of TOCTOU races, anyway. In other words, if anything breaks by this patch, then it's broken already.

## compat: there's no `<sys/time.h>` on MSVC

...but we won't be really using `gettimeofday()` there because `pthread_mutex_timedlock` is provided by the pthread compatibility library that we depend on.

## compat: make sure `ssize_t` is available

Normally this comes via `<sys/types.h>` on POSIX, but we're faking that on MSVC via our `<unistd.h>` shim. Make sure it's available.

## MSVC: compat: implement `localtime_r`

The Windows version of `localtime_s` swaps the argument order and has a different return type for increased level of fun.

## compat: MUSL implementation of `strptime()`

On Windows there's no `strptime()`. A similar functionality is available from C++11's `std::get_time` in `<iomanip>`, but there are subtle differences, including actual behavioral differences when parsing invalid dates such as "January 41st" (which is invalid everywhere, and could become "January 4th", eating one char less), or Feb 29th, 2018 (which does not exist).

Let's just import the MIT-licensed implementation from the MUSL C library because it's reasonably self-contained and appears to work the same as the glibc one.

## compat: strptime: lobotomize to not depend on `nl_langinfo`

...which is not available on Windows, and neither is `<strings.h>`.

## MSVC: compat: disable `pthread_mutex_timedlock` on Windows

On Windows there's no `gettimeofday()`/`nanosleep()`, and the usual suggestions come with a big disclaimer saying that it is not intended for high-resolution timing. Since this is not needed by the initial port of libyang to MSVC, let's just comment this out.

## MSVC: implement some of the atomic operations via MSVC' Interlocked ops

These actually look sane; the outliers are the load/store pairs, but given that the "alternative implementation" is just a plain old variable load...

## MSVC: compat: make the POSIX shims available to all targets

Some of them have been already defined, but others are being added via submodules. Best just make them available everywhere.

## MSVC: there's no `mmap()` on Windows

There's a compatibility library, the mman-win32, which implement a mmap-like semantics. However, on Windows one cannot map a read-only file into memory for a length that is larger than the size of that file, and you also cannot overwrite an existing mapping.

The easiest course of action is to just read the file into memory and be done with it.

See-also: https://stackoverflow.com/a/46014637/2245623

## MSVC: do not link with `-lm`

...because the Internet says that math functions are directly in MSVCR.

## MSVC: tools: disable interactive tools

I don't feel like porting `linenoise` to Windows, there's no `getline`, etc.

## MSVC: compat: `strcasecmp()` in `<strings.h>`

The `#ifdef` is for `_MSC_VER`, not `_WIN32`, because it's apparently available on MinGW.

## MSVC: do not build `fuzz/` and `style/` tests on Windows

Stuff like `waitpid` is a sure sign that `fuzz/` is better skipped when building on Windows. The `style/` is about compiling like C++ and about making sure that all files which need some functions include `compat.h`.  I don't feel like researching how to do `-Werror=c++-compat` on MSVC, and if the `compat.h` is ever missing, the MSVC build will yell loudly.

## MSVC: tests: link with all the compat libraries

This dance only has to be done because tests link against `yangobj` instead of simply using `-lyang` for linking against the library. I wonder why it's done like that.

## MSVC: tests: fix linking with `yangobj`

This fixes the LNK4217 linker warning. This project for some reasons creates an object library (`yangobjs` target in CMake) and proceeds to link tests against that instead of the usual way of linking against `libyang` the library properly. This is a problem on MSVC because the tests would normally process the API symbols as `__declspec(dllimport)`, while at the same time linking the object library with an implementation of these symbols into the same image. At the same time we cannot simply set `-DLIBYANG_BUILD` globally ebcause we're *also* building tools such as `yanglint` which does *not* embed these object files, but links against `-lyang` normally, and that one needs `__declspec(dllimport)`.

The alternative would be to get rid of the `yangobj` altogether, which would make the code more straightforward IMHO. That would also make the previous commit (`MSVC: tests: link with all the compat libraries`) superfluous because the usual rules of CMake transitive linking of libraries would take over.

See-also: https://docs.microsoft.com/en-us/cpp/error-messages/tool-errors/linker-tools-warning-lnk4217

## Windows: use binary I/O everywhere

...because newlines matter, at least in the test suite, and because having a stable file format really helps. No LF/CRLF changes for me, please.

## Windows: tests: adapt path handling

There's no exec bit on Windows, so don't test for that.

## Use semicolon as a directory delimiter on Windows

Windows path names have ':' inside as a drive letter separator. The widespread convention on that platform is to use ';' as a list separator in file paths, so let's follow suite.

It looks like this multi-value searchdir is only used in `ly_ctx_new()`, and that all other functions operate on one directory at a time.

## Windows: tests: do not invoke syscalls on an invalid FDs

On Windows, calling `fstat()` on a FD that was already closed results in calling the "invalid parameter handler". Set up on so that we can catch this properly. I'm doing this instead of just commenting out everything because I would prefer to keep that coverage to make sure we won't leak the FD in question.

## MSVC: tests: fix paths to utests and test plugins

On Unix, libyang plugins are not subject to `RUNTIME_OUTPUT_DIRECTORY`, instead requiring `LIBRARY_OUTPUT_DIRECTORY`. On Windows, the MSVC generator is a multi-config one, which means that it will happily put files into an extra subdirectory of `RUNTIME_OUTPUT_DIRECTORY`. This is super-hard to fix, so instead just query CMake for whatever the generator thinks is suitable, and use *that*.

This means that we have to distinguish between "the test binary directory", "directory with utests binaries" and "directory with test plugin binaries", so while we're at that, make sure that these paths are different than they used to be before this change -- simply to make sure that all bugs will be visible even on single-config generators like make or ninja. The tests still use the "binary dir" for things like `CMakeFiles` (!), which is super implementation-specific, but at least now the actual binaries live elsewhere.

Querying the `xxx_OUTPUT_DIRECTORY`, however, is not enough because that property does *not* have the config-specific suffix. Generators to the rescue, but that also means that the generated file suddenly has to be generator-config specific as well, which means that we have to massage the incldue path even more. Yay.

Also, on windows the test executables have the `.exe` suffix.

## MSVC: tests: use backslashes in expected normalized directory names

...and because `__FILE__` already has The Correct™ format even on Windows, use that for a non-existing file as well.

## tests: Plugins are not supported on Windows, yet

For some reason the `dlopen()` always fails, like this:

```
[ RUN      ] test_add_simple
libyang[2]: Searching for "ietf-yang-types" in "D:\a\oopt-gnpy-libyang\build\libyang\tests\utests".
libyang[2]: Newer revision than "ietf-yang-types@2013-07-15" not found, using this as the latest revision.
libyang[2]: Searching for "ietf-inet-types" in "D:\a\oopt-gnpy-libyang\build\libyang\tests\utests".
libyang[2]: Newer revision than "ietf-inet-types@2013-07-15" not found, using this as the latest revision.
libyang[0]: Loading "D:/a/oopt-gnpy-libyang/build/libyang/out-plugins/Release/plugin_simple.dll" as a plugin failed ("D:\a\oopt-gnpy-libyang\build\libyang\out-plugins\Release\plugin_simple.dll": The specified module could not be found.).
[  ERROR   ] --- 0x0 != 0x2
[   LINE   ] --- D:\a\oopt-gnpy-libyang\oopt-gnpy-libyang\libyang\tests\utests\basic\test_plugins.c:50: error: Failure!
[  FAILED  ] test_add_simple
```

Rather than trying to debug this on a platform that I don't really know, let's declare plugins as not supported on Windows.

## Allow building on Windows under MSVC as well

Legend of warnings:
- `/wd4711`, don't care about inline
- `/w14013`, missing prototype is a serious thing

## MSVC: source code is in UTF-8

This is important because the test suite uses `\Uxxxx` for some Unicode chars.

## CI: remove an unused job
    
This was used for the coverity job before, but that one got moved to another file.

Fixes: 7f55a30c actions REFACTOR separate devel push actions

## CI: start building on Windows

The differences compared to a POSIX build are too big even in the build script, mainly because CMake considers the Visual Studio build file generator a multiconfig one, and that means that all invocations of cmake/ctest have to have a specification of the current config. There's no way around that in CMake (and boy did I try!), so let's give up and do everything from scratch.

### CODE REVIEW

File: src/common.c
Comment (michalvasko) [inline src/common.c:424]: I don't particularly like such a long code in `#ifdef` directly in the main source file. Maybe you could create a Windows-specific source file with functions called from the main source.

File: tools/lint/CMakeLists.txt
Comment (michalvasko) [inline tools/lint/CMakeLists.txt:7]: This is adding a fair amount of complexity for `yanglint` compilation, I do not like this. But I am not sure how to improve it so I am let this one go.

File: compat/posix-shims/strings.h
Comment (michalvasko) [inline compat/posix-shims/strings.h:6]: Missing `#else`.

File: compat/compat.c
Comment (michalvasko) [inline compat/compat.c:305]: Add `#else`.

File: tests/CMakeLists.txt
Comment (michalvasko) [inline tests/CMakeLists.txt:17]: Comment for the commit adding this (saying that `yangobj` target should be removed altogether): I totally agree but the reason for it being written this way is that there are test for some functions in the internal *libyang* API. I was against this from the beginning (it was not my call then) and the latest plans were to fix this and rewrite the tests to use only public API but we did not make this change yet.

File: src/context.c
Comment (michalvasko) [inline src/context.c:261]: Could just be `PATH_SEPARATOR_STR` renamed to `PATH_SEPARATOR` and here `PATH_SEPARATOR[0]` used.

File: tests/utests/basic/test_inout.c
Comment (michalvasko) [inline tests/utests/basic/test_inout.c:58]: For all I care you can remove this test instead of this wild code.

File: tests/CMakeLists.txt
Comment (michalvasko) [inline tests/CMakeLists.txt:59]: `MSVC: tests: fix paths to utests and test plugins` This is another huge layer of complexity, why not just skip these tests instead?

File: tests/utests/basic/test_context.c
Comment (michalvasko) [inline tests/utests/basic/test_context.c:54]: `MSVC: tests: use backslashes in expected normalized directory names` Again, perhaps rather skip this test altogether.

File: src/plugins_types/ipv4_address.c
Comment (michalvasko) [inline src/plugins_types/ipv4_address.c:19]: General minor note, I like more having `#ifdef` instead of `#ifndef`.

File: CMakeLists.txt
Comment (michalvasko) [inline CMakeLists.txt:5]: Don't really like this but I suppose if the newer version is strictly required not much to do about it.

**Review by michalvasko — CHANGES_REQUESTED** [review]

File: src/build.h
Comment (jktjkt) [reply src/build.h:32]: I realize that this is a "noisy change", but then that's why I've done that in a single commit over all of libyang. It's just the way that platform works; all symbols that are exported to a DLL must be marked as such by "method A" in the public header, and then by the same "method A" in the implementation, and then when user code includes these symbols, the first "method A" must become "method B". I tried to explain that in the commit message, but you're right that this needs docs.

I'll be happy to move it to any other header you indicate. Which one should I use? The only constrain is that it must be included by all public headers *and* by all `.c` files which implement any of these public symbols. Once I know the location, I'll add the docs as well.

File: compat/compat.c
Comment (jktjkt) [reply compat/compat.c:265]: Thanks, I'll change all implementations in `compat.c` (that's `realpath`, `localtime_r`, `dirname`, `setenv`).

File: src/build.h
Comment (michalvasko) [reply src/build.h:32]: Maybe use `config.h`? Also, when adding some comments, please rename `LIBYANG_API_IMPL` to `LIBYANG_API_DEF` (definition) and `LIBYANG_API` to `LIBYANG_API_DECL` (declaration), it is more clear to me.

File: tools/lint/CMakeLists.txt
Comment (jktjkt) [reply tools/lint/CMakeLists.txt:7]: Thanks. The alternative would be bringing in "something like `linenoise`", but for Windows, and I do not really prefer that :).

File: src/build.h
Comment (jktjkt) [reply src/build.h]: Done.

File: CMakeLists.txt
Comment (jktjkt) [reply CMakeLists.txt:5]: Agreed, which is why it's only for Windows. Some of the fancier target tricks are definitely a no-go on CMake 2.8.12, and since this is a new port and I can only test on the latest, it makes sense to use that IMHO.

File: tests/CMakeLists.txt
Comment (jktjkt) [reply tests/CMakeLists.txt:17]: that makes sens,e I've adjusted the commit message.

File: src/common.c
Comment (jktjkt) [reply src/common.c]: Done.

File: tests/utests/basic/test_inout.c
Comment (jktjkt) [reply tests/utests/basic/test_inout.c]: It was there to check that the FD is not leaked, but I agree, that's a super-minor thing. Gone.

File: src/build.h
Comment (jktjkt) [reply src/build.h]: Just FYI, this requested change broke compilation under Windows due to the way how their headers are structured. I'll solve that via bringing the `compat.h` earlier in the include chain in a later commit (this is once again related to their [dreaded winsock implementation](https://stackoverflow.com/questions/1372480/c-redefinition-header-files-winsock2-h)).

File: tests/CMakeLists.txt
Comment (jktjkt) [reply tests/CMakeLists.txt:59]: OK. I've added a trivial test for Windows to make sure that everything still works (this helped me before to catch that `:` path separator).

File: tests/utests/basic/test_context.c
Comment (jktjkt) [reply tests/utests/basic/test_context.c:54]: Simplified now.

**Comment by jktjkt** [comment]
OK, I think that all feedback has been incorporated now, and I made one final `uncrustify` sweep to make the CI happy (the changes appear minor and also touch `strptime.c` which is a third-party import, so I chose to make that sweep in an extra commit at the end).

**Comment by jktjkt** [comment]
I don't think that the "NetDEF-CI" failures are specific to this patch series (e.g., the `yangre` manpage doesn't look relevant). Should I pay attention to them?

**Comment by michalvasko** [comment]
No, it is fine, they are not relevant.

File: compat/faux_mmap.c
Comment (michalvasko) [inline compat/faux_mmap.c:14]: `compat` is meant to have only implementations of standard functions, this is *libyang* specific (the `common.h` include). Sorry for not mentioning any details, please either put these functions in a separate `common_win.c` or something like that to `libyang/src` (if you anticipate more functions may be added there later) or put these functions directly in `common.c`. I do not mind having `#ifdef WIN32` with a single function call in another Linux function, I am trying to avoid having large Windows-specific code block in a Linux function, that is all. So you may keep `ly_mmap` as it was and just call Windows code as needed with code `ly_win_mmap();`, for example.

File: tools/lint/main_ni_only.c
Comment (michalvasko) [inline tools/lint/main_ni_only.c:4]: Update doxygen metadata.

File: CMakeLists.txt
Comment (michalvasko) [inline CMakeLists.txt:73]: Please switch to `if(WIN32)`.

**Review by michalvasko — CHANGES_REQUESTED** [review]
Mostly minor thing left, do not worry if I do not merge this today, I will take a look and finish this tomorrow.

File: compat/faux_mmap.c
Comment (jktjkt) [reply compat/faux_mmap.c]: OK, moved back to `common.c`. The reason why I have not done it like that in my first series is because there's some overlap in the leading code of `ly_mmap` and I want to use that.

**Comment by jktjkt** [comment]
> I made one final `uncrustify` sweep to make the CI happy

I've now separated out all the required changes to appropriate commits and checked that there's no coding style issue at any point (except for these two `strptime` commits which are fixed right after that). In order to make this checkable automatically, I had to fix an error in CMake setup (another `make` assumption while I'm on `ninja`) -- so there's a new commit right at the beginning of this patch series.

**Review by michalvasko — APPROVED** [review]

**Comment by michalvasko** [comment]
Alright, thanks for the contribution, it will hopefully be helpful for more users.

**Pull request merged**
